### ISSUE TITLE
USB scale not working on Mac

### ISSUE THREAD

**Issue by klabarge** [qzind/qz-print#180]
Mac OS X El Capitan 10.11.3 
Chrome, Firefox, and Safari
I am using a Stamps.com Model 510 that works successfully on Windows

I am able to **Claim Device** and **Open Stream** properly:
1. After **Open Stream** is selected, the **Read Data** screen is displayed, and no values are read off of the scale.  Nothing is displayed in the console at this point.
   ![image](https://cloud.githubusercontent.com/assets/12505463/13340619/c74f7056-dbff-11e5-9c3e-c17d1ca6a4f5.png)
2. When clicking **"Read as Weight"** the Raw data dialogue does not change and the below console errors are shown.  This error indefinitely happens until **Read as Raw** is selected.
   ##### Chrome Console
   
   ![image](https://cloud.githubusercontent.com/assets/12505463/13340696/66c6ae1a-dc00-11e5-954a-82f2704a2b8e.png)
   ##### Firefox Console
   
   ![image](https://cloud.githubusercontent.com/assets/12505463/13341105/a39ad6e2-dc03-11e5-9eb9-50368abe0977.png)

**Comment by tresf** [comment]
I'm impressed you could claim the device... I'm getting this in an Apple OS X 10.11 VM:

```
Error: USB error 3: Unable to claim interface: Access denied (insufficient permissions)
```

Seems to be a duplicate of [`stackoverflow/28884817`](http://stackoverflow.com/questions/28884817/usb4java-library-error-while-claiming-an-interface) which states OSX claims the devices preventing other applications from doing so.  The work-arounds are tragically ugly (creating dummy `kext` kernel extensions for every `VendorID`, `ProductID` that needs to be made available, reboot).

**Comment by klabarge** [comment]
I initially received the Access Denied error as well.  I'm not exactly sure how I got around this, as I did nothing besides basic troubleshooting such as restarting the browser and QZ Tray.

I would like to note that I am receiving almost the exact error on Ubuntu 14.04

##### Ubuntu 14.04

```
Error: USB error 3: Can't open device Bus 001 Device 013: ID 1446:6a73: Access denied (insufficient permissions)
```

**Comment by tresf** [comment]
It appears Apple fixed some of the USB issues with software updates and my VM is behind... Updating now... 

Possible upstream bug report https://github.com/libusb/libusb/issues/96.

**Comment by tresf** [comment]
More information about this bug... https://github.com/libusb/libusb/wiki/FAQ#How_can_I_run_libusb_applications_under_Mac_OS_X_if_there_is_already_a_kernel_extension_installed_for_the_device

**Edit:** Kext signing request submitted to Apple.  Follow-up: 636837176
**Edit2:** Started via https://github.com/qzind/qz-print/compare/2.0...tresf:2.0

**Comment by tresf** [comment]
I think I got it via https://github.com/qzind/qz-print/compare/2.0...tresf:2.0

Note, until Apple approves my request for developer signing rights for kernel extensions (`kext`), the following bash must be run prior to installing.  
1. Install this temporary installer [`qz-tray-2.0.0-unsigned-kext3.pkg`](https://github.com/tresf/qz-print/releases/download/v2.0.0-RC2/qz-tray-2.0.0-unsigned-kext3.pkg)
2. Run the following command:
   
   ``` bash
   sudo nvram boot-args=kext-dev-mode=1
   ```
3. Then boot in recovery mode and disable System Integrity Protection:
   1. Reboot Mac
   2. Hold the following keys: `Command` + `R`
   3. When prompted for Recovery, go to `Utilities, Terminal`
   4. Type:
      
      ``` bash
      csrutil disable
      reboot
      ```
4. Wait for OSX to load up again...
5. Fire up QZ, sample.html
6. Try to read scale

Notes:
- The reboot step is mandatory even after I get signing rights but only needs to happen after the first install if USB support is needed.
- It didn't work my first attempt, but it may have been a problem with the USB pass-thru with VMWare fusion.
- @bberenz there seems to be a range error going on as can be seen in the below picture.  The Raw data seems to alternate between `03,04,0c,fe,af,00,00,00` and `04,04,00,00,00,00,00,00`.

![screen shot 2016-02-26 at 2 35 35 am](https://cloud.githubusercontent.com/assets/6345473/13346200/a266d9d6-dc32-11e5-9a1a-6ba527044a05.png)

**Comment by klabarge** [comment]
After following the above steps, I now get a `NullPointerException` when trying to "List Device Interfaces."

Even after rebooting a second time, I get the same error.

![image](https://cloud.githubusercontent.com/assets/12505463/13368620/fb61dd92-dcb9-11e5-8d46-9ae87d2cdbbd.png)

**Comment by tresf** [comment]
Can you you start in debug mode and tell is where the `NullPointerException` is being fired from?

Also, despite that, can you still claim the device?

**Comment by klabarge** [comment]
I cannot claim the device.  I get "Identifier cannot be empty."
- Here's the stack trace when I click "List Device Interfaces"

```
[DEBUG] 2016-02-26 21:39:57,824 @ qz.ws.PrintSocketClient:?
    Message: {"call":"usb.listInterfaces","promise":{},"params":{"vendorId":"0x1446","productId":"0x6a73"},"timestamp":1456540797823,"uid":"wat88r"}
[WARN] 2016-02-26 21:39:57,824 @ qz.ws.PrintSocketClient:?
    Bad signature on request
[INFO] 2016-02-26 21:39:58,934 @ qz.common.TrayManager:?
    Allowed localhost to access USB devices
[ERROR] 2016-02-26 21:39:58,935 @ qz.ws.PrintSocketClient:?
    Problem processing message
java.lang.NullPointerException
    at qz.utils.UsbUtilities.getDeviceInterfaces(Unknown Source)
    at qz.utils.UsbUtilities.getDeviceInterfacesJSON(Unknown Source)
    at qz.ws.PrintSocketClient.processMessage(Unknown Source)
    at qz.ws.PrintSocketClient.onMessage(Unknown Source)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
    at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
    at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
    at java.lang.Thread.run(Thread.java:745)
```
- Here's the stack trace when I try to claim the device while inputting the known Device Interface

```
[DEBUG] 2016-02-26 21:41:46,274 @ qz.ws.PrintSocketClient:?
    Message: {"call":"usb.claimDevice","promise":{},"params":{"vendorId":"0x1446","productId":"0x6a73","interface":"0x00"},"timestamp":1456540906273,"uid":"tw4y9f"}
[WARN] 2016-02-26 21:41:46,274 @ qz.ws.PrintSocketClient:?
    Bad signature on request
[INFO] 2016-02-26 21:41:47,428 @ qz.common.TrayManager:?
    Allowed localhost to claim a USB device
[ERROR] 2016-02-26 21:41:47,429 @ qz.ws.PrintSocketClient:?
    Problem processing message
java.lang.NullPointerException
    at qz.communication.UsbIO.open(Unknown Source)
    at qz.ws.PrintSocketClient.processMessage(Unknown Source)
    at qz.ws.PrintSocketClient.onMessage(Unknown Source)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70)
    at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:68)
    at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver$2.run(JettyAnnotatedEventDriver.java:210)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by tresf** [comment]
@bberenz I'm going to need some help to fix the `NullPointerException`.

I've narrowed it down to the fact that this USB device on Mac for some reason has no "Active" configuration.

The work-around for this was to setup a helper function to grab the active configuration and if `null`, fallback to the **first, non-null** configuration.  This successfully prevents the buttons `List Device Interfaces` and `List Interface Endpoints` from returning a NPE.

``` java
public static UsbConfiguration getUsbConfiguration(UsbDevice device) throws UsbException {
     UsbConfiguration config = device.getActiveUsbConfiguration();
     if (config != null) {
         return config;
     }

     for (Object o : device.getUsbConfigurations()) {
         if (o != null) {
             return (UsbConfiguration)o;
         }
     }

     throw new UsbException("UsbConfiguration not found for vendorId/productId");
}
```

Per: https://github.com/tresf/qz-print/commit/d53236e95d56fa2990d30728b086999131144ca0#diff-66e51979c4dd4f537b85c3203a0f1568R120

However, a strict mandate of the USB library is to only allow the claiming of an active configuration.  This means when we click `Claim Device`, a new exception is thrown `Configuration not active`.

My attempts to workaround this have all failed, but mostly due to my lack of understanding of the way this library works.

The JSR says you can call `StandardRequest.setInterface(...)` to set the configuration as active, but I'm not sure what the parameters should be (or even if this could be accomplished easier using a libusb4java call.

``` diff
public void open(byte ifc) throws UsbException {
-     iface = device.getActiveUsbConfiguration().getUsbInterface(ifc);
+     iface = UsbUtilities.getUsbConfiguration(device).getUsbInterface(ifc);
+
+     // FIXME
+     // Sets the interface index, alternate settings index incase the interface isn't active
+     //if (!iface.isActive()) {
+     //    StandardRequest.setInterface(device, ifc, (byte)0x00);
+     //}
+
       iface.claim();
}
```

Per: https://github.com/tresf/qz-print/commit/d53236e95d56fa2990d30728b086999131144ca0#diff-52ae2363b650928540f4715b3eae938fR36

Help appreciated.

**Comment by tresf** [comment]
Turns out the `nvram` hack died hard with an OS X update, so the unsigned kernel extension work-around had to be amended.  Instructions have been updated.  https://github.com/qzind/qz-print/issues/180#issuecomment-189151458.  Please perform these with caution as they disable integrity checks for the OS.

New installer link has been provided as well.

Note, this still suffers the `Configuration not active` bug (https://github.com/qzind/qz-print/issues/180#issuecomment-189741530), which for some reason I'm now able to reproduce on my 10.11 VM.

Research on https://github.com/qzind/qz-print/issues/180#issuecomment-189741530 suggests [`usb_set_configuration()` is actually recommended on OS X](http://libusb.6.n5.nabble.com/Mac-OS-X-configuration-not-set-td4384677.html) for libusb, just a matter of finding out the High-Level `javax.usb` api which calls it...

**Comment by tresf** [comment]
Still banging my head on this... Filed upstream bug report to ask for help.https://github.com/usb4java/usb4java/issues/51

**Comment by tresf** [comment]
Low-level API didn't help at all.

After re-reading some `libusb` threads again, I'm pretty sure the OS X fix hasn't made it downstream yet (10.11.4) per https://github.com/libusb/libusb/issues/123.

I signed up for the OS X beta program.... Installing now...

![image](https://cloud.githubusercontent.com/assets/6345473/13386930/b61aae4a-de7f-11e5-9135-29caf3276d99.png)

**Comment by tresf** [comment]
**Update:** 10.11.4 still does NOT fix this issue.  Edit filed an upstream bug report over at `libusb`. https://github.com/libusb/libusb/issues/158

**Comment by tresf** [comment]
Well, I made every valiant attempt possible to no avail.  Closing as `wont-fix`, `not-our-bug`.

The long history of trial and error is better documented with the `libusb` devs over at libusb/libusb#158.

Here's the last working kernel extension build if we ever have to go back on it... https://github.com/tresf/qz-print/commit/f70c1a5dc27526d7274542671c89a3e8e6ba9399

This feature will be added using official HID support when `2.1` is released.  Until then, no USB scale support on Apple (and we can remove it from our test scripts for now).

**Comment by tresf** [comment]
As a status update on this, `kext` signing **was approved** for our Apple Developer ID.

We're unlikely to use it for claiming HID devices on Mac, but it's available in case we encounter hardware that requires it.

```
Team ID: P5DMU6659X 
Company / Organization: QZ INDUSTRIES LLC
Company / Product URL: https://qz.io
Intended Distribution: Consumer or Enterprise Distribution
Reason for Request: Allow a list of pre-approved USB devices through libusb.
```

Update:  HID support has been started in the `tray` repo here:  https://github.com/qzind/tray/commit/1d2f230722c9cfc4b5723418d9aa97f159f4c188, available sometime after the 2.0.0-stable release.
